📖 Add MachinePool controller architecture docs

CecileRobertMichon opened the pull request:
<!-- please add a icon to the title of this PR (see https://sigs.k8s.io/cluster-api/VERSIONING.md), and delete this line and similar ones -->
<!-- the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, feature additions), 🐛 (:bug:, patch and bugfixes), 📖 (:book:, documentation or proposals), or 🌱 (:seedling:, minor or other) -->

**What this PR does / why we need it**: Adds controller architecture docs for MachinePools in the CAPI book.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3244

CecileRobertMichon commented:
/assign @devigned

devigned commented on docs/book/src/developer/architecture/controllers/machine-pool.md:
```diff
@@ -0,0 +1,116 @@
+
+#### Required `spec` fields
+
+The `spec` object **must** at least one field defined:
```
```suggestion
The `spec` object **must** have at least one field defined:
```

devigned commented on docs/book/src/developer/architecture/controllers/machine-pool.md:
```diff
@@ -0,0 +1,116 @@
+
+#### Required `status` fields
+
+The `status` object **must** at least one field defined:
```
```suggestion
The `status` object **must** have at least one field defined:
```

devigned reviewed (changes requested):
Looks good. Just a couple little changes.

fabriziopandini commented:
Thanks for this documentation!
/milestone v0.4.0
/lgtm

Wondering if it worth to backport this given that this applies to the 0.3 branches as well...

CecileRobertMichon commented:
> Wondering if it worth to backport this given that this applies to the 0.3 branches as well...

 I don't think a backport is necessary given that MachinePool is in exp/ in v1alpha3. Doing this for v1alpha4 as a pre-requisite to move MachinePool out of exp.

CecileRobertMichon commented:
/assign @vincepri

wfernandes commented on docs/book/src/developer/architecture/controllers/machine-pool.md:
```diff
@@ -0,0 +1,114 @@
+
+The Machine controller's main responsibilities are:
+
+* Setting an OwnerReference on:
+    * Each MachinePool object to the Cluster object.
+    * The associated BootstrapConfig object.
+    * The associated InfrastructureMachinePool object.
```
```suggestion
* Setting an OwnerReference on each MachinePool object to:
    * The associated Cluster object.
    * The associated BootstrapConfig object.
    * The associated InfrastructureMachinePool object.
```

wfernandes commented on docs/book/src/developer/architecture/controllers/machine-pool.md:
```diff
@@ -0,0 +1,114 @@
+`MachinePool.Spec.Template.Spec.Bootstrap.DataSecretName` is empty.
+* Setting NodeRefs to be able to associate instances and kubernetes nodes.
+* Deleting Nodes in the target cluster when the associated instance is deleted.
+* Cleanup of related objects.
```
What are the related objects? Is it referring to things like BootstrapConfig?

CecileRobertMichon replied:
removed, this doesn't apply here. There is one bootstrap config per machine pool (not one per instance)

wfernandes commented on docs/book/src/developer/architecture/controllers/machine-pool.md:
```diff
@@ -0,0 +1,114 @@
+    * The associated InfrastructureMachinePool object.
+* Copy data from `BootstrapConfig.Status.DataSecretName` to `MachinePool.Spec.Template.Spec.Bootstrap.DataSecretName` if
+`MachinePool.Spec.Template.Spec.Bootstrap.DataSecretName` is empty.
+* Setting NodeRefs to be able to associate instances and kubernetes nodes.
```
Are these two the same? Just wanted to clarify where the NodeRefs are being set and what instances were we referring to.
```suggestion
* Setting NodeRefs on MachinePool instances to be able to associate them with kubernetes nodes.
```

wfernandes commented:
I want to educate myself on MachinePools since I haven't interacted much with them. 🙂 
/assign

wfernandes commented on docs/book/src/developer/architecture/controllers/machine-pool.md:
```diff
@@ -0,0 +1,114 @@
+* Copy data from `BootstrapConfig.Status.DataSecretName` to `MachinePool.Spec.Template.Spec.Bootstrap.DataSecretName` if
+`MachinePool.Spec.Template.Spec.Bootstrap.DataSecretName` is empty.
+* Setting NodeRefs to be able to associate instances and kubernetes nodes.
+* Deleting Nodes in the target cluster when the associated instance is deleted.
```
nit: Again just clarifying.
```suggestion
* Deleting Nodes in the target cluster when the associated MachinePool instance is deleted.
```

wfernandes commented on docs/book/src/developer/architecture/controllers/machine-pool.md:
```diff
@@ -0,0 +1,114 @@
+
+### Infrastructure provider
+
+The InfrastructureMachine object **must** have both `spec` and `status` objects.
```
Should this be InfrastructureMachinePool? 
For example, I noticed that the `AWSMachinePoolSpec` has `ProviderIDList` whereas the `AWSMachineSpec` only has a `ProviderID` field.
```suggestion
The InfrastructureMachinePool object **must** have both `spec` and `status` objects.
```

wfernandes commented on docs/book/src/developer/architecture/controllers/machine-pool.md:
```diff
@@ -0,0 +1,114 @@
+
+The `spec` object **must** have at least one field defined:
+
+* `providerIDList` - the list of cloud provider IDs identifying the instances.
```
> identifying the instances.

Are these machine instances or node instances? I'm guessing they're the same.

CecileRobertMichon replied:
they're provider instances, for example ASG instances or VMSS instances. 1 instance = 1 node

wfernandes commented on docs/book/src/developer/architecture/controllers/machine-pool.md:
```diff
@@ -0,0 +1,114 @@
+
+| secret name | field name | content |
+|:---:|:---:|---|
+|`<cluster-name>-kubeconfig`|`value`|base64 encoded kubeconfig that is authenticated with the child cluster|
```
On line 14 we have _target cluster_,
on line 17,24 we have _workload cluster_.

Suggesting we use _workload cluster_ everywhere since that's what we have in the glossary.

```suggestion
|`<cluster-name>-kubeconfig`|`value`|base64 encoded kubeconfig that is authenticated with the workload cluster|
```

wfernandes reviewed (commented):
So when I run `make serve-book`, I don't get a generated html file for machine-pool, in the architecture section. However, after I added a reference in the SUMMARY.md (which we should probably do as well), then the html was generated.

wfernandes commented:
Also, does this mean MachinePool is now being graduated? If so we should probably remove, https://cluster-api.sigs.k8s.io/tasks/experimental-features/machine-pools.html. 
If not yet, then it may be helpful to put a link in the above doc to the MachinePool controllers page. WDYT?

CecileRobertMichon commented on docs/book/src/developer/architecture/controllers/machine-pool.md:
```diff
@@ -0,0 +1,114 @@
+# Machine  Controller
```
whoops, good catch

CecileRobertMichon commented:
Thanks for the review @wfernandes! Added the new doc to SUMMARY.md.

> Also, does this mean MachinePool is now being graduated? If so we should probably remove, https://cluster-api.sigs.k8s.io/tasks/experimental-features/machine-pools.html.
If not yet, then it may be helpful to put a link in the above doc to the MachinePool controllers page. WDYT?

Not quite yet, this is one of the preparation steps in order to graduate it. Added a link the exp doc.

devigned reviewed (approved):
lgtm

wfernandes commented:
/lgtm

vincepri reviewed (approved):
/approve
